ingestion-beam/src/main/java/com/mozilla/telemetry/io/Write.java

```diff
@@ -148,6 +168,109 @@ public FileOutput(ValueProvider<String> outputPrefix, OutputFileFormat format,
     }
   }
 
+  /**
+   * Implementation of writing to local or remote files.
+   *
+   * <p>For details of the intended behavior for file paths, see:
+   * https://github.com/mozilla/gcp-ingestion/tree/master/ingestion-beam#output-path-specification
+   */
+  public static class AvroOutput extends Write {
+
+    private final ValueProvider<String> outputPrefix;
+    private final Duration windowDuration;
+    private final ValueProvider<Integer> numShards;
+    private final Compression compression;
+    private final InputType inputType;
+    private final AvroSchemaStore schemaStore;
+
+    /** Public constructor. */
+    public AvroOutput(ValueProvider<String> outputPrefix, Duration windowDuration,
+        ValueProvider<Integer> numShards, Compression compression, InputType inputType,
+        ValueProvider<String> schemasLocation) {
+      this.outputPrefix = outputPrefix;
+      this.windowDuration = windowDuration;
+      this.numShards = numShards;
+      this.compression = compression;
+      this.inputType = inputType;
+      this.schemaStore = AvroSchemaStore.of(schemasLocation);
+    }
+
+    /** A functor that constructs a GenericRecord from a PubsubMessage. */
+    public class RecordFormatter implements AvroIO.RecordFormatter<PubsubMessage> {
+
+      /** Convert a PubsubMessage into an Avro record with the provided schema. */
+      public GenericRecord formatRecord(PubsubMessage element, Schema schema) {
+        String message;
+        GenericRecord result;
+        try {
+          message = Json.asString(element);
+          DatumReader<GenericRecord> reader = new GenericDatumReader<>(schema);
+          Decoder decoder = DecoderFactory.get().jsonDecoder(schema, message);
+          result = reader.read(null, decoder);
+        } catch (IOException e) {
+          throw new UncheckedIOException(e);
+        } catch (AvroTypeException e) {
+          throw e;
```

Comment:
This breaks the pipeline during tests. I'd like to propagate this up with something like a `UncheckedRuntimeException`.

Reply:
It may be a limitation of `sinkViaGenericRecords` that it doesn't allow any facility for handling exceptions raised while formatting records. We might need to be responsible for doing more of that checking ourselves.

Reply:
I've solved this by creating a DoFn that handles side-input and tuple-tags for a success and error PCollection.